File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
	"net"
	"strings"
)

const (
	MsgTypeBatch uint16 = 1
	MsgTypeAck   uint16 = 2
	MsgTypeError uint16 = 3

	TypeBytes   = 2
	LengthBytes = 4

	PayloadDelimiter = ","
	BatchDelimiter   = "\n"
)

// SendMessage Constructs and sends a message with the given type and data to the server. It returns an error if there was an issue sending the message.
func SendMessage(conn net.Conn, msgType uint16, data [][]string) error {
	if msgType != MsgTypeBatch {
		return fmt.Errorf("Invalid message type: %d", msgType)
	}

	parts := make([]string, len(data))
	for i, d := range data {
		parts[i] = strings.Join(d, PayloadDelimiter)
	}
	payload := []byte(strings.Join(parts, BatchDelimiter))

	msgTypeHeader := make([]byte, TypeBytes)
	binary.BigEndian.PutUint16(msgTypeHeader, msgType)

	lengthHeader := make([]byte, LengthBytes)
	binary.BigEndian.PutUint32(lengthHeader, uint32(len(payload)))

	if err := sendAll(conn, msgTypeHeader); err != nil {
		return err
	}

	if err := sendAll(conn, lengthHeader); err != nil {
		return err
	}

	if len(payload) > 0 {
		if err := sendAll(conn, payload); err != nil {
			return err
		}
	}

	return nil
}

// ReceiveMessage Reads a message from the server, returning the message type, payload, and any error encountered.
func ReceiveMessage(conn net.Conn) (uint16, []string, error) {
	msgTypeHeader := make([]byte, TypeBytes)
	if err := receiveAll(conn, msgTypeHeader); err != nil {
		return 0, nil, err
	}
	msgType := binary.BigEndian.Uint16(msgTypeHeader)
	if msgType != MsgTypeAck && msgType != MsgTypeError {
		return 0, nil, fmt.Errorf("Invalid message type: %d", msgType)
	}

	lengthHeader := make([]byte, LengthBytes)
	if err := receiveAll(conn, lengthHeader); err != nil {
		return 0, nil, err
	}
	length := binary.BigEndian.Uint32(lengthHeader)

	if length == 0 {
		return msgType, []string{}, nil
	}

	payload := make([]byte, length)
	if err := receiveAll(conn, payload); err != nil {
		return 0, nil, err
	}

	return msgType, strings.Split(string(payload), PayloadDelimiter), nil
}

// sendAll Is a helper to avoid short writes. It ensures that all bytes in the data slice are sent over the connection, returning an error if any write operation fails.
func sendAll(conn net.Conn, data []byte) error {
	totalSent := 0
	for totalSent < len(data) {
		sent, err := conn.Write(data[totalSent:])
		if err != nil {
			return err
		}
		totalSent += sent
	}
	return nil
}

// receiveAll Is a helper to avoid short reads. It ensures that all bytes in the buf slice are read from the connection, returning an error if any read operation fails.
func receiveAll(conn net.Conn, buf []byte) error {
	totalRead := 0
	for totalRead < len(buf) {
		read, err := conn.Read(buf[totalRead:])
		if err != nil {
			return err
		}
		totalRead += read
	}
	return nil
}

File: server/common/protocol.py
import socket
from .utils import Bet


class ConnectionClosedError(Exception):
    pass


class ProtocolError(Exception):
    pass


class Protocol:
    MSG_TYPE_BET = 1
    MSG_TYPE_ACK = 2

    TYPE_BYTES = 2
    LENGTH_BYTES = 4

    PAYLOAD_DELIMITER = ","

    @staticmethod
    def recv_all(sock: socket, expected_length: int) -> bytes:
        """
        Helper to avoid short reads.
        It ensures that all bytes in the buf slice are read from the connection, returning an error if any read operation fails.
        """
        data = bytearray()
        while len(data) < expected_length:
            packet = sock.recv(expected_length - len(data))
            if not packet:
                raise ConnectionClosedError("Connection closed while receiving data")
            data.extend(packet)
        return data

    @staticmethod
    def receive_message(sock: socket) -> tuple[int, list[str]]:
        """
        Reads a message from the socket, returning a tuple of (msgtype, payload).
        """
        msgtype_header = Protocol.recv_all(sock, Protocol.TYPE_BYTES)
        msgtype = int.from_bytes(msgtype_header, byteorder="big", signed=False)
        if msgtype != Protocol.MSG_TYPE_BET:
            raise ProtocolError(f"Invalid message type: {msgtype}")

        length_header = Protocol.recv_all(sock, Protocol.LENGTH_BYTES)
        length = int.from_bytes(length_header, byteorder="big", signed=False)

        if length == 0:
            return msgtype, []

        payload = Protocol.recv_all(sock, length)
        data = payload.decode("utf-8")
        return msgtype, data.split(Protocol.PAYLOAD_DELIMITER)

    @staticmethod
    def send_message(sock: socket, msgtype: int, data: list[str]) -> None:
        """
        Construct and sends a message with the given type and data to the socket.
        """
        if msgtype != Protocol.MSG_TYPE_ACK:
            raise ProtocolError(f"Invalid message type: {msgtype}")

        payload = Protocol.PAYLOAD_DELIMITER.join(data).encode("utf-8")

        msg_type_header = msgtype.to_bytes(
            Protocol.TYPE_BYTES, byteorder="big", signed=False
        )

        length_header = len(payload).to_bytes(
            Protocol.LENGTH_BYTES, byteorder="big", signed=False
        )

        sock.sendall(msg_type_header + length_header)

        if payload:
            sock.sendall(payload)
